core/src/main/java/org/mule/runtime/core/privileged/processor/CompositeProcessorChainRouter.java

```diff
@@ -7,79 +7,94 @@
 import static org.mule.runtime.core.api.lifecycle.LifecycleUtils.stopIfNeeded;
+import static org.mule.runtime.core.api.processor.MessageProcessors.processWithChildContext;
+import static org.mule.runtime.core.internal.event.DefaultEventContext.child;
+import static org.slf4j.LoggerFactory.getLogger;
+import static reactor.core.publisher.Flux.fromIterable;
 import static reactor.core.publisher.Mono.from;
+
 import org.mule.runtime.api.exception.MuleException;
 import org.mule.runtime.api.lifecycle.InitialisationException;
 import org.mule.runtime.api.lifecycle.Lifecycle;
 import org.mule.runtime.core.api.InternalEvent;
+import org.mule.runtime.core.api.InternalEventContext;
 import org.mule.runtime.core.api.MuleContext;
 import org.mule.runtime.core.api.processor.MessageProcessorChain;
 import org.mule.runtime.core.api.processor.ReactiveProcessor;
 import org.mule.runtime.core.privileged.component.AbstractExecutableComponent;
-import org.mule.runtime.core.privileged.processor.chain.DefaultMessageProcessorChainBuilder;
 
 import java.util.List;
-import java.util.function.Function;
+import java.util.function.BiFunction;
 
 import javax.inject.Inject;
 
-import org.reactivestreams.Publisher;
 import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 /**
- * Component to be used that supports a collection of {@link MessageProcessorChain} and executes them in order.
+ * Component to be used that supports a collection of {@link MessageProcessorChain} and executes them in sequentially waiting for
+ * each chain and all child context executions (such as async blocks) to complete.
  * <p/>
  * Meant to be used by runtime privileged extensions that need to construct top level chains of execution.
  * 
  * @since 4.0
  */
 public class CompositeProcessorChainRouter extends AbstractExecutableComponent implements Lifecycle {
 
-  private static Logger LOGGER = LoggerFactory.getLogger(CompositeProcessorChainRouter.class);
+  private static Logger LOGGER = getLogger(CompositeProcessorChainRouter.class);
 
   @Inject
   private MuleContext muleContext;
 
-  private String name;
   private List<MessageProcessorChain> processorChains = emptyList();
-  private MessageProcessorChain messageProcessorChain;
 
-  public void setProcessorChains(List processorChains) {
+  public void setProcessorChains(List<MessageProcessorChain> processorChains) {
     this.processorChains = processorChains;
   }
 
   @Override
   protected ReactiveProcessor getExecutableFunction() {
-    return publisher -> from(publisher).transform(messageProcessorChain);
+    return publisher -> from(publisher).flatMapMany(initial -> fromIterable(processorChains).reduce(initial, processChain()));
+  }
+
+  private BiFunction<InternalEvent, MessageProcessorChain, InternalEvent> processChain() {
+    return (event, processorChain) -> {
+      InternalEventContext childContext = child(event.getContext(), ofNullable(getLocation()));
```

Comment:
shouldn't the location of the scatter gather be the location of the component this router?